keras/engine/training.py

```diff
@@ -3769,18 +3771,25 @@ def reduce_per_replica(values, strategy, reduction="auto"):
     or a `tf.Tensor`, if the strategy has already conducted the reduction
     for the downstream library.
 
-    There are three possible outcomes of reduction:
+    There are five possible outcomes of reduction:
 
     1) if the `values` is a structure of simple `tf.Tensor`s, meaning that
        reduction is not actually needed, `reduce_per_replica` returns the
        structure as-is.
-    2) else, if `reduction="first"`, then `reduce_per_replica`
+    2) else, if `reduction="auto"`, then it assumes "first" if running
```

Comment:
I think we should either not mention the details of choosing the reduction methods, or explain why it's chosen this way (for which, I'm afraid there's not an easy way to describe in docs as it really depends on tf.distribute strategies' implementation).

So, I would probably just say the library would choose the best reduction method based on the training environment, for 2).